Docs: Update monitoring documentation for grafana and prometheus (#324)
codacy/chart: docs/configuration/monitoring.md

# Monitoring

Currently two monitoring solutions are supported:

* [**Crow**](##Setting-up-monitoring-using-Crow) - A simple, lightweight, and builtin monitoring solution.

> * [**Prometheus+Grafana+Loki**](##Setting-up-monitoring-using-Grafana,-Prometheus,-and-Loki) - A more comprehensive third-party monitoring solution, for more advanced usage.

Reviewer: Fix links and minor edits:

```suggestion
* [**Crow**](#setting-up-monitoring-using-crow) - A simple, lightweight, and built-in monitoring solution.
* [**Prometheus + Grafana + Loki**](#setting-up-monitoring-using-grafana-prometheus-and-loki) - A comprehensive third-party monitoring solution, recommended for more advanced usage.
```

## Setting up monitoring using Crow

Crow is a visualization tool that displays information about the projects and jobs

> that are pending for analysis, as well as the ones running in the Codacy platform.

Reviewer: Just to clarify, I interpret this as "Crow displays information about the **projects pending analysis** and the **jobs currently running** on Codacy". Is this correct?

Author: I barely touched the Crow documentation. I just moved it to the top of the document. It was changed just a week ago so I assumed it was thoroughly reviewed then. These sentences seem equivalent, but I can change it nonetheless.

The Crow tool is installed alongside Codacy after the helm chart is deployed to the cluster.
It can be accessed through the `/monitoring` path of the url pointing to your Codacy
installation, e.g. `http://codacy.company.org/monitoring`. You must set the
`global.codacy.crow.url` value in your `values.yaml` file so that anchor links to your
projects can be properly established inside `crow`.

For example:

```yaml
global:
  codacy:
    crow:
      url: "http://codacy.example.com/monitoring"
```

> Please see the [README.md](https://github.com/codacy/chart/blob/master/README.md) for more information about these values.

Reviewer: I think we can drop this detail, as we are not mentioning the values reference in the `README.md` for the other .yaml values files.

Advanced users who may want to change the values besides the ones we identify with placeholders will probably know where to find the information, no?

Author: As above. Crow isn't really for advanced users, but I also see no need in this line. I'll remove it.

### Configuring your credentials

> You **should** provide a password for the `crow` installation either through the `values.yaml` file or through a `--set` parameter during the `helm` installation process. This parameter can be configured as follows:

Reviewer: Defining the custom password is not mandatory (since there is a default one) but highly recommended, is this right? Also, to improve the documentation readability we can simply use the capitalized proper nouns of the tools throughout this page, when appropriate of course:

```suggestion
We highly recommend that you define a custom password for the Crow installation. You can do this either through the `values.yaml` file or through a `--set` parameter during the Helm installation process. This parameter can be configured as follows:
```

Author: "Should" is generally used to convey a recommendation or an expected behaviour, and doesn't imply something is mandatory.

* Through a `--set` parameter:

```yaml
helm upgrade (...) --set crow.config.passwordAuth.password=<--- crow password --->
```

* Through the `values.yaml` file:

```yaml
crow:
  config:
    passwordAuth:
      password: <--- crow password --->
```

**If you have not configured your `crow` credentials as described above
the following default credentials will be used:**

```yaml
username: codacy
password: C0dacy123
```

## Setting up monitoring using Grafana, Prometheus, and Loki

[Prometheus](https://prometheus.io) is an open-source systems monitoring and alerting
toolkit. Logs can be collected using [Loki](https://grafana.com/oss/loki/), which is a
horizontally-scalable, highly-available, multi-tenant log aggregation system
It's data can be visualized with [Grafana](https://grafana.com), a widely used

> open source analytics and monitoring solution.

Reviewer: ```suggestion
Its data can be visualized with [Grafana](https://grafana.com), a widely used
open-source analytics and monitoring solution.
```

Author: Good catch for the typo in the possessive. Thanks. Regarding `open source` I note that the authorities on the subject, e.g. the  [Open Source Initiative](https://opensource.org), do not hyphenate.

The following guide covers the basic installation of the components. This monitoring
stack is considerably more resource demanding than Crow, and is recommended only for
more advanced usage. Furthermore, its installation, configuration, and management
requires a deeper knowledge of Kubernetes as each component must be carefully tweaked

> to match your specific use case, using as starting point the values files herein provided.

Reviewer: ```suggestion
The following guide covers the basic installation of the components in this monitoring stack.
This solution is considerably more resource demanding than Crow, and is recommended only for
more advanced usage. Furthermore, its installation, configuration, and management
requires a deeper knowledge of Kubernetes as each component must be carefully tweaked
to match your specific use case, using as starting point the .yaml values files provided by us.
```

Author: Herein is more informative because, while it implies that the files are provided by us, it also states that they are in this repo and not elsewhere. The rest is somewhat equivalent, but I don't mind changing it.

> ### Installing CRDs

Reviewer: What are CRDs? :nerd_face: I genuinely don't know, and there's a chance that customers may not know as well.

Author: They are [Custom Resource Definitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/), i.e. extensions of the Kubernetes API. I surely hope that people are not installing CRDs without knowing what they are, and as we stated above, this stack is for more advanced users. But I don't mind adding a link to the documentation.

The simplest way to setup prometheus in your cluster is by using the
[Prometheus-Operator](https://github.com/helm/charts/tree/master/stable/prometheus-operator)
bundle. Start by adding the custom resources required for installing this bundle in your cluster.

> NOTE: if installing on MicroK8s use `microk8s.kubectl` instead of `kubectl`

Reviewer: ```suggestion
**NOTE:** if installing on MicroK8s use `microk8s.kubectl` instead of `kubectl`
```

Reviewer: @paulopontesm The note instructing users to run `microk8s.kubectl` is getting spread out throughout the documentation. Maybe it's better to recommend using the alias in the MicroK8s-specific quick guide? (This would be something for a different pull request, though.)

kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.36/example/prometheus-operator-crd/monitoring.coreos.com_prometheuses.yaml
kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.36/example/prometheus-operator-crd/monitoring.coreos.com_prometheusrules.yaml
kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.36/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml
kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.36/example/prometheus-operator-crd/monitoring.coreos.com_thanosrulers.yaml
```

### Install Loki

Obtain the configuration file for Loki, [values-loki.yaml](https://github.com/codacy/chart/blob/master/codacy/values-loki.yaml),
and install it as described below. You may need to edit the storage class for Loki
persistence in this configuration file. While the default setting should suit most use
cases, you may need to adjust it to your specific Kubernetes installation.

> For instance, for MicroK8s use `storageClassName: microk8s-hostpath`.

Reviewer: ```suggestion
Obtain the configuration file for Loki, [values-loki.yaml](https://github.com/codacy/chart/blob/master/codacy/values-loki.yaml),
and install it as described below. While the default storage class setting
for Loki persistence should suit most use cases, you may need to adjust
it to your specific Kubernetes installation. For instance, for MicroK8s use
`storageClassName: microk8s-hostpath`.
```

kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.36/example/prometheus-operator-crd/monitoring.coreos.com_prometheuses.yaml
kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.36/example/prometheus-operator-crd/monitoring.coreos.com_prometheusrules.yaml
kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.36/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml
kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.36/example/prometheus-operator-crd/monitoring.coreos.com_thanosrulers.yaml
```

### Install Loki

Obtain the configuration file for Loki, [values-loki.yaml](https://github.com/codacy/chart/blob/master/codacy/values-loki.yaml),
and install it as described below. You may need to edit the storage class for Loki
persistence in this configuration file. While the default setting should suit most use
cases, you may need to adjust it to your specific Kubernetes installation.
For instance, for MicroK8s use `storageClassName: microk8s-hostpath`.

```bash
helm repo add loki https://grafana.github.io/loki/charts

> kubectl create namespace monitoring

Reviewer 2: ```suggestion
```

Reviewer 2: I think this was meant to be deleted 🤔

Author: Done. Thanks. Forgot that one :)

The simplest way to setup prometheus in your cluster is by using the
[Prometheus-Operator](https://github.com/helm/charts/tree/master/stable/prometheus-operator)

> bundle. Start by adding the custom resources required for installing this bundle in your cluster.

Reviewer 2: I think we need to separate documents for the "crow" way of monitoring from the Prometheus + Loki + Grafana.

Also.. I believe it would be more clear if we just went through each of the components one by one:

# 1. Install Prometheus

> One liner saying what is prometheus

> A few line on how to install

# 2. Install Loki

# 3. Install Grafana

> Including how to configure the ingress to access it....

# 4. Enable Codacy ServiceMonitors and Dashboards

Author: Yes, I'll make it more legible. Note that Grafana and prometheus are installed by the [same chart](https://github.com/helm/charts/tree/master/stable/prometheus-operator).

Given that the documentation for monitoring is currently not vast it is probably not too hard to follow. Having the docs for crow and prometheus+loki+grafana in the same place has the virtue of making it easier for a user to get info about the monitoring options available.

We could make it clearer by changing the text at the top of the document, and possibly inverting the order, so that the simpler and built-in solution (crow) is documented first. I'll tweak it a bit.

```sh

> # NOTE: if installing on microk8s use `microk8s.kubectl` instead of `kubectl`

Reviewer: I would place the *NOTE:* outside the code block because it will be more visible. Later I will try to find a way to format all note / warning / ... text so it is more user-friendly and looks nicer.  :)

kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.36/example/prometheus-operator-crd/monitoring.coreos.com_prometheuses.yaml
kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.36/example/prometheus-operator-crd/monitoring.coreos.com_prometheusrules.yaml
kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.36/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml
kubectl apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/release-0.36/example/prometheus-operator-crd/monitoring.coreos.com_thanosrulers.yaml
```

Obtain the configurations files for [loki](monitoring-conf/loki-values.yaml),
[promtail](monitoring-conf/promtail-values.yaml), and
[prometheus-operator](monitoring-conf/prometheus-operator-values.yaml) and place them
in the same folder where you'll run the helm commands.

**You must edit:**

-   the Grafana password for the `admin` user in `prometheus-operator-values.yaml`
-   the storage class for loki persistence in `loki-values.yaml`

You can now install the monitoring bundle on your cluster:

> ```sh

Reviewer: `sh` works just as fine as `bash`, but for the sake of consistency I suggest that we keep using "bash" throughout like in all other Markdown files in this repository.

helm repo add loki https://grafana.github.io/loki/charts

kubectl create namespace monitoring

helm upgrade --install --atomic loki loki/loki --version 0.17.0 \
  --recreate-pods --namespace monitoring --values loki-values.yaml

helm upgrade --install --atomic promtail loki/promtail --version 0.13.0 \
  --recreate-pods --namespace monitoring --values promtail-values.yaml

helm upgrade --install --atomic monitoring stable/prometheus-operator --version 6.9.3 \
  --recreate-pods --namespace monitoring --values prometheus-operator-values.yaml
```

> Now that you have prometheus and grafana installed you can enable `serviceMonitors`

Reviewer: Just a nitpick, but check the capitalization of the 3rd party tools whenever is makes sense (when we're not referring to a lowercase command or setting, for example):

```suggestion
Now that you have Prometheus and Grafana installed you can enable `serviceMonitors`
```

      enabled: true
```

Apply this configuration by issuing an helm upgrade while passing these additional values.
To do so append `--values monitoring-values.yaml --recreate-pods` to the command used to install
codacy provided in [this documentation](../install.md).

```sh
helm upgrade (...options used to install codacy...) --values monitoring-values.yaml --recreate-pods
```

> Grafana will be running in the same domain as Codacy, on port 30080, for instance

Reviewer 2: How does this happen? We are not configuring the ingress for this, right?

Author: Yeah, it was a nodeport, but that's messy. I'll remove it for now, and then use codacy ingress when we add it to the chart.

You **should** provide a password for the `crow` installation either through the `values.yaml` file or through a `--set` parameter during the `helm` installation process. This parameter can be configured as follows:

> -   Through a `--set` parameter:

Reviewer 2: You need to rebase.. there is a new config for remark that forces `*` instead of `-`

Author: That's strage. I'm based on `ca2d5bb`, "Sandbox/slack artifact url", and there was only commit after that, which didn't change remark. Codacy also didn't pickup anything. But I'll check.

      enabled: true
```

Apply this configuration by issuing an helm upgrade while passing these additional values.
To do so append `--values values-monitoring.yaml --recreate-pods` to the command used to install
codacy provided in [this documentation](../index.md#2-installing-codacy).

```bash

> helm upgrade (...options used to install codacy...) --values values-monitoring-values.yaml --recreate-pods

Reviewer: We have a similar `helm upgrade` command above for the Crow credentials but we aren't explaining how the user should complete the command like we do here. Should we choose to provide more detailed instructions like here, or less like above? :thinking:

Author: I'd rather add more info to the command above. `(...)` alone isn't often used to document command line tools, and its meaning might be obscure.
